File: 0647 - Palindromic Substrings/solution.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size(), res = 0, l = 0, r = 0, i, j;
        while(r < n){
            while(r<n && s[l]==s[r]) r++;
            res += (r-l)*(r-l+1)/2;
            i = l-1, j = r;
            while(i>=0 && j<n && s[i]==s[j]) i--, j++, res++;
            l = r;
        }
        return res;
    }
};

File: 0647 - Palindromic Substrings/solution.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        res, n = 0, len(s)
        dp = [[False]*n for _ in range(n)]
        for i in range(n-1, -1, -1):
            for j in range(i, n):
                dp[i][j] = s[i]==s[j] and (j-i<3 or dp[i+1][j-1])
                res += dp[i][j]
        return res